src/WebJobs.Script.Grpc/Channel/OrderedInvocationMessageDispatcher.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Script.Grpc.Eventing;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.Grpc;
+
+internal sealed class OrderedInvocationMessageDispatcher : IInvocationMessageDispatcher, IDisposable
+{
+    private readonly ILogger _logger;
+    private readonly string _invocationId;
+    private readonly Action<InboundGrpcEvent> _dispatchWithChannel;
+
+    // Separated these out for easier testing of the flow.
+    private readonly Action<InboundGrpcEvent> _dispatchWithThreadPool;
+
+    private Channel<InboundGrpcEvent> _channel;
+    private bool _isChannelInitialized = false;
+    private bool _invocationComplete = false;
+    private bool _disposed = false;
+
+    public OrderedInvocationMessageDispatcher(string invocationId, ILogger logger, Action<InboundGrpcEvent> processItem)
+        : this(invocationId, logger, processItem, processItem)
+    {
+    }
+
+    // For testing
+    internal OrderedInvocationMessageDispatcher(string invocationId, ILogger logger, Action<InboundGrpcEvent> processItemWithChannel,
+        Action<InboundGrpcEvent> processItemWithThreadPool)
+    {
+        _logger = logger;
+        _invocationId = invocationId;
+        _dispatchWithChannel = processItemWithChannel;
+        _dispatchWithThreadPool = processItemWithThreadPool;
+    }
+
+    // For testing
+    internal Channel<InboundGrpcEvent> MessageChannel => _channel;
+
+    private static Channel<InboundGrpcEvent> InitializeChannel() =>
+        Channel.CreateUnbounded<InboundGrpcEvent>(
+            new UnboundedChannelOptions
+            {
+                SingleReader = true,
+                SingleWriter = true
+            });
+
+    public void DispatchRpcLog(InboundGrpcEvent msg)
+    {
+        // The channel is only needed if we use any logs. Otherwise, ordering doesn't matter.
+        // This is not thread-safe, but it is always called in-order from the same thread, so no
+        // need for locking.
```

Comment:
Is there any harm in locking? Could we end up in a situation where the host is stretched for resources and we end up with a race or the thread misordering the request somehow? If there is a way for us to be more defensive here I would like to take that approach

Reply:
Speed... we're trying to maximize throughput in this flow. But because everything gets dispatched here via the single-threaded grpc message handler, we can reliably know that at this point, it cannot be out-of-order. Once we get off this thread, then all bets are off, though.

Reply:
I guess a better way to put it -- this class is only meant to be used in this very specific scenario (called from single thread, etc). I've now updated comments and docs to make this clear. Any changes here should be inspected thoroughly in the future, but this is not a general-purpose class and any changes to the invocation pipeline in the future will need a bunch of testing and review. So I'm not too worried about accidentally calling it from multiple threads.

I also suspect that this will be removed or re-thought as we redo the worker pipeline in the near future. So it may not be long for this world...

Reply:
I didn't answer your question directly, I just realized...

> Could we end up in a situation where the host is stretched for resources and we end up with a race or the thread misordering the request somehow?

Not unless we significantly change how we process incoming grpc messages. Today, from one worker they all go to a single `System.Threading.Channel`, which has a single consumer whose job is to dequeue a message and "dispatch" them off that one thread for processing. Today (before this change) -- that is here -- `_inbound` is the `System.Threading.ChannelReader` for that worker:

https://github.com/Azure/azure-functions-host/blob/b0e811adac38d7334fb5ca6a58531bb7bfdeb61e/src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs#L246-L256

It's a loop whose job is to read a single message, "dispatch" it, and then get the next one. So, there's no way for it to dispatch anything from multiple threads. It cannot move to another one without handling the previous. If we were to move to a multiple reader scenario, we'd have to re-think a bunch of stuff -- so we wouldn't do that without some significant investigation and review.